### DISCUSSION TITLE
Issue With Naming TV Episodes

### DISCUSSION THREAD

**Discussion by jonathanmorlen** [automatic-ripping-machine/automatic-ripping-machine#724] [Q&A] [answered]
When I start a job, the show is able to be located, but the episodes themselves aren't able to be identified from the IFO file on the disk or OMDB API or wherever it originates. This results in multiple A1_t00.mkv files being generated and me having to manually rename each to avoid collisions. Where in the logs is the episode name attempting to be found?

Thanks!

**Comment by 1337-server** [comment, marked as answer]
Its not attempted at all, as this info isnt provided from the disc. or OMDB.

At least not that i am aware of ? Im happy to be wrong though. And would love to see auto episode recognition added. But as far as I'm aware neither HandBrake or MakeMKV are able to retrieve this info, and OMDB only retrieves the correct title, year, poster image and plot.

**Reply by josh-hemphill** [reply]
Might it be more useful then to use the ARM disk database or another database to maintain a list of known disk->episodes mappings?

**Reply by microtechno9000** [reply]
We are working on a major rewrite with ARM at the moment to solve this and other issues

**Reply by josh-hemphill** [reply]
Awesome!
Is there anything I can do to help? Is it that big PR draft?

**Reply by microtechno9000** [reply]
Always keen for support from users, and your welcome to help with coding if you can. 

The major PR is part of the Development branch update, aiming to seperate the UI and ripper. It also introduces MySQL.

If you are familiar with docker and building from scratch, welcome pulling and testing by users. Alpha testing is always good, just note at the moment there are known bugs and missing components.

**Reply by mezion1** [reply]
I am likewise dealing with series. My current solution to collision is to add disc number into name. The issue I was having was wanting to set file name - e.g. Set job to "Show Name S01D01" for /Show Name S01D01/Show Name S01D01_Track01.mkv - and then I should be able to ensure relative uniqueness, just requiring me to updating disc and season numbers as required. I can then dump them into parent folders more easily.
To achieve this, I updated /opt/arm/ripper/handbrake.py line 206 and 209 (searched for "UPDATED filename:"), and inserted job.title into the output file name.
`track.filename = job.title + "_" + destfile + "." + cfg.arm_config["DEST_EXT"]`
and 
`filename = os.path.join(basepath, job.title + "_" + destfile + "." + cfg.arm_config["DEST_EXT"])`
I tried to figure out how to set it in makemkv, but as far as I could see, makemkv only provides options (via .MakeMKV/settings.conf) against extracted track "Title" information, which my test discs tracks did not have set - hence I get things like C1_t01. Maybe useful for setting a name on a per rip basis, but no good for multiple rips in automation jobs. Windows version appeared to require an overwrite (cut disc title, click out, paste back) of title to apply the title naming rules to the subsequent tracks. It seems a common question online? I read it as requiring rewriting/adding title record to track info after the track info is read, which I could not see how to do when using makemkvcon -r. 
Anyway, it was the solution I found for series file naming. It is really late, I hope this will be useful to others

**Comment by ghost** [comment]
If there's no automatic way to do this, could there be a way to manually input it?
like have a page that lists the episode files and the user can select what the season and episode number is and enter a name that would be added to the file name once transcoding is completed. then they can just look at the back of the dvd box to determine which episodes are on each disk

**Reply by mihawk90** [reply]
> then they can just look at the back of the dvd box to determine which episodes are on each disk

The problem with this approach is that the "titles" on disc are not necessarily in the order the episodes were aired or are on the backcover. The way Blurays are authored means that the video files that come out of MakeMKV don't have to be in order. When you hit play in the menu this corresponds to a playlist playing video files (or rather fragments) in a set order, no matter where they are on disc.

Case in point I was recently ripping my [House](https://www.themoviedb.org/tv/1408-house) Blurays and while for *most* of the discs they are in order, I had 2 or 3 discs where that was not the case.
So for example the titles extracted by MakeMKV might be 1, 2, 3, 4, 5, but they correspond to episodes 5, 2, 3, 1, 4 for example. There is likely *some* way to rip them in the "right" order, but that is something that would have to be addressed in MakeMKV, there's nothing ARM can do about it.

TLDR you cannot rely on the disc layout to identify episodes, you need to check the actual video file to properly map them.

**Reply by ghost** [reply]
Has anyone made an online database that contains disk names, the shows they're associated with, and what the order on that particular disk is (and the first scene that can be used to identify the specific episode?)

**Reply by mihawk90** [reply]
Not that I know of. Using the title for identification is likely also not a good idea since it's not guaranteed to be unique. MakeMKV itself uses a hash of one of the files on disc for unique identification for decryption purposes (sorry forgot which one), which is probably the way to go.
As for episode mappings it's also not quite as easy. Some discs have the same episode under different playlists (why-ever that would be the case I don't know), and MakeMKV seems to pick whichever it finds first at the time of the initial scan as the "main" title, so it's not completely consistent either. Can of course be solved with allowing multiple playlist mappings to an episode.

**Reply by judokan9** [reply]
@Betonhaus
When it comes to the order of the episodes, the segment assignment and the name of the source file should help. In 99% of cases, the order of the episodes would then be correct. See [https://forum.makemkv.com/forum/viewtopic.php?f=6&t=20327](https://forum.makemkv.com/forum/viewtopic.php?f=6&t=20327).

@1337-server
Are there any plans for an option to adjust the naming accordingly (mpls or segment mapping) so that series episodes can be numbered correctly and burned-in subtitles can be better identified (like in Star Wars) as explained in the attached link?

**Reply by mihawk90** [reply]
> @Betonhaus
> When it comes to the order of the episodes, the segment assignment and the name of the source file should help. In 99% of cases, the order of the episodes would then be correct. See https://forum.makemkv.com/forum/viewtopic.php?f=6&t=20327.

This is exactly what I was talking about above, not every disc is authored this way (another example is mentioned in that very thread).
Yes this will work in *most* cases but it is by no means guaranteed.

**Reply by judokan9** [reply]
> This is exactly what I was talking about above, not every disc is authored this way

Ahh, I see!
It has worked very well for me so far and is a hundred times better than the standard naming.
So it would be cool, in my eyes, to have it as an option that the files get the mpls names.
In any case, I'm with you that it's a pity there's no reliable standard way to find out the episode order.
But the workaround from above comes very close with a few exceptions.

**Reply by mihawk90** [reply]
> to have it as an option that the files get the mpls names

You can already do that. See [the Wiki entry for MakeMKV](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/MakeMKV-Info#settingsconf) and the links provided there.

**Reply by judokan9** [reply]
Ahh, I had only used the `app_DefaultSelectionString` until now.
Using `app_DefaultOutputFileName` I can now also output the sourceid.

Thanks @mihawk90  I did not know that this possibility exists.

For anyone who would also like to have the `sourceid` in their names:
I have added the following entry in the `settings.conf` which is located in the `.MakeMKV` folder:
```
app_DefaultOutputFileName = "{NAME1}{-:CMNT1}{-:DT}{title:+DFLT}{-sourceid-:SN}{_t:N2}"
```
Which extends the default naming schema by `{-sourceid-:SN}` and thus creates names like:

`MySeries Disc 1-sourceid-00800_t00.mkv`

And can thus make the naming of episodes in series easier.

**Reply by mihawk90** [reply]
There's also a thing called [pHashes](https://www.phash.org) (perceptual hashing) that can be used to identify the *contents* of the video even across different resolutions and qualities too (roughly what AcoustID is for audio on musicbrainz). In essence it is what Google Images uses to search for "similar" images (although their results are often... "not great", but that comes down to how "fuzzy" one makes the search).

But to my knowledge there is no database for movies/shows using this, and even if there were, I think that'd be something for a media manager rather then a ripper anyway.

**Reply by josh-hemphill** [reply]
Isn't that what ARM already does for movies? Maintaining a DB for identifying movies so it can automatically fill in the name and id?
Doing the same thing TV sounded like it's going to require a new DB service anyway, so it might be worth creating one that uses pHashes and can be extended to also migrate the movie identification to later.

**Reply by mihawk90** [reply]
> Isn't that what ARM already does for movies? Maintaining a DB for identifying movies so it can automatically fill in the name and id?

I don't actually know TBH, I never looked into the DB stuff. I know in the WebUI there is a "Send DVDs to API" but I don't know whether a) that even still works and b) if it's also used for Blurays. That part of ARM doesn't seem to be documented at all.

Right now I know it's using the disc titles, and I always assumed it's just searching that on OMDB when enabled (but I don't use the OMDB lookup).

> Doing the same thing TV sounded like it's going to require a new DB service anyway, so it might be worth creating one that uses pHashes and can be extended to also migrate the movie identification to later.

I'm sure it's technically possible, but is it really advisable? ARM is supposed to be a tool for ripping discs, not for media management. There are already plenty of tools that deal with media library management, so if anything it should be integrated there.
Feature creep is a real thing and that sounds very much like one, especially with only not-even-a-handful devs.

**Reply by josh-hemphill** [reply]
> I'm sure it's technically possible, but is it really advisable? ARM is supposed to be a tool for ripping discs, not for media management. There are already plenty of tools that deal with media library management, so if anything it should be integrated there.
Feature creep is a real thing and that sounds very much like one, especially with only not-even-a-handful devs.

Except that it's already doing it, you can use it to just rip, but it also does the identification/media-conversion/organization in a second step if you choose. If there was some software that could handle everything else that ARM does, that'd be great, but right now I use the direct output of ARM's identification and conversion steps to be able to dump onto my jellyfin server, and as far as I know, jellyfin/plex/emby expect all your media to already be identified, tagged, and organized; that way they have the most certainty about what your providing it.

**Reply by mihawk90** [reply]
> Except that it's already doing it, you can use it to just rip, but it also does the identification/media-conversion/organization in a second step if you choose.

There's a difference between supporting existing code and introducing an entire new code path with an entire new database, a database which would have to exist in the first place.

**Reply by SylvainMT** [reply]
> Isn't that what ARM already does for movies? Maintaining a DB for identifying movies so it can automatically fill in the name and id?

ARM doesn't do this at all.  It queries your selected Metadata Provider (either OMDB, the default or TMDB) and attempts to find the title of the movie there and apply that to the disk you are ripping.  ARM itself, knows nothing of the contents of the disk, and if you did not provide an API key to your MetaData provider it will never find any title, other then the title of the disk.

**Reply by banshee** [reply]
> I'm sure it's technically possible, but is it really advisable? ARM is supposed to be a tool for ripping discs, not for media management.

It's not just ripping though - it's pretty well integrated with things like HandBrake.  If you're pulling in something like a stack of 4k blurays, it's hard to see how to avoid doing media management in ARM itself, since you probably want quite a bit of the raw information from the disk to do the identification.  Once you've encoded the files and thrown away the originals, it becomes MUCH harder to figure out how to catalog the contents.

**Reply by banshee** [reply]
> Once you've encoded the files and thrown away the originals, it becomes MUCH harder to figure out how to catalog the contents.

So, I just spent a few minutes seeing how this is done, and I'm wildly wrong.  Part of why I'm wildly wrong is that I did work on this kind of stuff professionally at mp3.com years ago.  We did checksums of various parts of the disk to get guaranteed exact matches for the catalog, and I just assumed that ARM/OMDB/TMDB worked in the same way.  Turns out that's not even close; it's just text matches.

**Reply by mihawk90** [reply]
> Once you've encoded the files and thrown away the originals, it becomes MUCH harder to figure out how to catalog the contents.

Not really. I already mentioned a way to do this that works across encodes and resolutions. There is already software that does this, but it requires a central database to do the hash-to-title mapping, which to my knowledge doesn't exist for TV shows and movies. 

> Turns out that's not even close; it's just text matches.

Yes that's what I mentioned above.
